### PR TITLE
Add friends filter for ranking

### LINKED ISSUE #2556: Filter friends on rankings is missing
Old site link:
https://osu.ppy.sh/p/pp

New site link:
https://osu.ppy.sh/rankings/osu/performance

### PR DESCRIPTION
**Pull request by notbakaneko**
could use some layout suggestions 🤔 @arflyte 

image

Resolves #2556.

### CODE REVIEW

**Comment by arflyte** [comment]
Just filter for friends, or there'll be more filter in the future?

**Comment by nanaya** [comment]
There's also an existing different filter

**Comment by notbakaneko** [comment]
image

**Comment by arflyte** [comment]
Figma_200409-133902

How about something like this?

**Comment by notbakaneko** [comment]
does country default to your country? where does it show which country is it being filtered by?

**Comment by nanaya** [comment]
Hi @arflyte I think the detail for how country filter works is missing as @notbakaneko mentioned.

**Comment by arflyte** [comment]
The filter "country" is actually more like "local". If they wanna see ranking of other countries they have to use the Country section and click on a country.

I'll think of a better way to make this less confusing.

**Comment by notbakaneko** [comment]
The country ranking and filter by country aren't the same;
This is country ranking:
image

This is filtering ranking by country. You can pick any country:
image

image

**Comment by nanaya** [comment]
Hi @arflyte and see above &uarr;. Maybe at least how the filter (or page) looks like when a specific country is selected.

**Comment by notbakaneko** [comment]
Also keep in mind that filtering by country only applies to performance ranking.

**Comment by arflyte** [comment]
Figma_200416-152249

I think it made more sense this way. Stay in the Country tab when countries are clicked and show a breadcrumb under the tabs.

**Comment by notbakaneko** [comment]
so where does the friends only filter fit in?_?

**Comment by arflyte** [comment]
Maybe there's no real quick way to solve this. I think I have to redesign the filter and the tabs. Let me try again.

**Comment by arflyte** [comment]
Figma_200416-175508
Figma_200416-175549

Aite, how's this?

**Comment by notbakaneko** [comment]
@arflyte so any many countries are going to be in that dropdown? All of them?

**Comment by arflyte** [comment]
Yeah. I think that's pretty much the same with other forms that list all the country. Unless we put the top 20 countries on the top of the list, then other countries in alphabetical order?

File: resources/assets/lib/ranking-filter.tsx [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+        bn='ranking-select-options'
+        renderItem={this.renderItem}
+        onItemSelected={this.handleItemSelected}
+        options={this.options}
+        selected={this.selectedCountry}
+      />
+    );
+  }
+
+  renderItem(item: RenderProps) {
```
Comment (nanaya) [inline resources/assets/lib/ranking-filter.tsx:98]: also unclear what is Item here

File: resources/assets/lib/ranking-filter.tsx [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+
+interface State {
+  country: string | null;
+}
+
+const allCountries = { id: null, text: osu.trans('rankings.countries.all') };
+
+export default class RankingFilter extends React.PureComponent<Props, State> {
+  static defaultProps = {
+    sortMode: 'all',
```
Comment (nanaya) [inline resources/assets/lib/ranking-filter.tsx:24]: it's not "sort", is it?

File: resources/assets/lib/ranking-filter.tsx [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+  };
+
+  readonly state: State = {
+    country: new URL(window.location.href).searchParams.get('country'),
```
[the remark this answers is not in the document]
Comment (notbakaneko) [reply resources/assets/lib/ranking-filter.tsx:28]: result of aborted attempt of moving the result tables to react 🙃

File: resources/assets/lib/ranking-filter.tsx [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+        </div>
+      </div>
+    );
+  }
+
+  renderCountries() {
+    if (this.props.type !== 'performance') return null;
+
+    return (
+      <SelectOptions
```
[the remark this answers is not in the document]
Comment (notbakaneko) [reply resources/assets/lib/ranking-filter.tsx:106]: https://github.com/ppy/osu-web/blob/bb5134a/resources/assets/less/variables.less#L102-L104 🤔

File: resources/assets/lib/ranking-filter.tsx [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+        bn='ranking-select-options'
+        renderItem={this.renderOption}
```
[the remark this answers is not in the document]
Comment (notbakaneko) [reply resources/assets/lib/ranking-filter.tsx:108]: All these renaming of existing component props sounds completely out of scope for this PR
